ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/NumberOfBarsCriterion.java

```diff
@@ -34,21 +37,25 @@ this software and associated documentation files (the "Software"), to deal in
 public class NumberOfBarsCriterion extends AbstractAnalysisCriterion {
 
     @Override
-    public double calculate(TimeSeries series, TradingRecord tradingRecord) {
-        int nBars = 0;
+    public Num calculate(TimeSeries series, TradingRecord tradingRecord) {
+        Num nBars = series.numOf(0);
         for (Trade trade : tradingRecord.getTrades()) {
-            nBars += calculate(series, trade);
+            nBars = nBars.plus(calculate(series, trade));
         }
         return nBars;
     }
 
     @Override
-    public double calculate(TimeSeries series, Trade trade) {
-        return (1 + trade.getExit().getIndex()) - trade.getEntry().getIndex();
+    public Num calculate(TimeSeries series, Trade trade) {
+        Objects.requireNonNull(series, "Series must be not null");
```

Comment:
NumberOfBars looks good.  Still wondering about dealing with `nulls` more consistently..

Reply:
Yeah me too, I'm doubting if it's really worth it in the end.

The `Objects.requireNonNull(series, "Series must be not null");` is only to have a 'better readable' error message, but it's basically the same as not doing it, bar the extra message.

```java
java.lang.NullPointerException at 
org.ta4j.core.analysis.criteria.NumberOfBarsCriterion.calculate(NumberOfBarsCriterion.java:41)
```
Would be the error without the error message, so in that sense it *is* better readable. But it's at a cost of being more bloated and how far do you go?

Take this line in the calculation for example:

```java
            (series.numOf(1).plus(series.numOf(trade.getExit().getIndex())))
```

What if the trade is only created using this constructor:

```java
public Trade(OrderType startingType)
``` 

In that case there is no exit order, so we'd still get a null pointer exception.

So to 'properly' do the null checks in this method we'd need to do:

```java
    @Override
    public Num calculate(TimeSeries series, Trade trade) {
        Objects.requireNonNull(series, "Series must be not null");
        Objects.requireNonNull(trade, "Trade must be not null");
        Objects.requireNonNull(trade.getExit(), "Trade's Exit Order must be not null");
        Objects.requireNonNull(trade.getEntry(), "Trade's Entry Order must be not null");

        return
            (series.numOf(1).plus(series.numOf(trade.getExit().getIndex())))
            .minus(series.numOf(trade.getEntry().getIndex()));
    }
```

And then i'm not even talking about any criterion that's doing for loops over the `TradeRecord`. 

Maybe we should just trust on the user of the Criterions to pass the correct data, and if they don't, they'll just have to handle the NullPointerExceptions...

Reply:
Another possibility is this legacy logic in `MaximumDrawdownCriterion`:
```
    @Override
    public double calculate(TimeSeries series, Trade trade) {
        if (trade != null && trade.getEntry() != null && trade.getExit() != null) {
            CashFlow cashFlow = new CashFlow(series, trade);
            Num maximumDrawdown = calculateMaximumDrawdown(series, cashFlow);
            return maximumDrawdown.doubleValue();
        }
        return 0;
    }
```
But that just "hides" the null parameters and returns an actual `double`.  Very misleading!

I suggest we just leave all of these as they currently are and open a new issue to standardize handling of criteria parameters regarding `nulls`.

Reply:
OOS, opened #216 .